### PR TITLE
refactor(core): Extract poll context out of NodeExecutionFunctions (no-changelog)

### PR DESCRIPTION
**Pull request by netroy**
## Summary
This PR extracts Poll context object out of NodeExecutionFunctions, and adds unit tests for most of this code.

## Related Linear tickets, Github issues, and Community forum posts

CAT-276

## Review / Merge checklist

- [x] PR title and summary are descriptive. ([conventions](../blob/master/.github/pull_request_title_conventions.md)) 
- [ ] [Docs updated](https://github.com/n8n-io/n8n-docs) or follow-up ticket created.
- [x] Tests included. 
- [ ] PR Labeled with `release/backport` (if the PR is an urgent fix that needs to be backported)

### CODE REVIEW

File: packages/core/src/NodeExecuteFunctions.ts [hunk 1]
```diff
@@ -167,6 +167,8 @@ import {
 import { extractValue } from './ExtractValue';
 import { InstanceSettings } from './InstanceSettings';
 import type { ExtendedValidationResult, IResponseError } from './Interfaces';
+// eslint-disable-next-line import/no-cycle
```
Comment (tomi) [inline packages/core/src/NodeExecuteFunctions.ts:170]: Could we move the functions `PollContext` needs from this file to a separate file so we wouldn't have to introduce a cycle?

File: packages/core/src/node-execution-context/base-contexts.ts [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+	NodeTypeAndVersion,
+	Workflow,
+	WorkflowExecuteMode,
+} from 'n8n-workflow';
+import { deepCopy, LoggerProxy } from 'n8n-workflow';
+import { Container } from 'typedi';
+
+import { InstanceSettings } from '@/InstanceSettings';
+
+export abstract class BaseContext implements Omit<FunctionsBase, 'getCredentials'> {
```
Comment (tomi) [inline packages/core/src/node-execution-context/base-contexts.ts:15]: Could we have documentation for this class

File: packages/core/src/node-execution-context/poll-context.ts [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+
+const throwOnEmit = () => {
+	throw new ApplicationError('Overwrite PollContext.__emit function');
+};
+
+const throwOnEmitError = () => {
+	throw new ApplicationError('Overwrite PollContext.__emitError function');
+};
+
+export class PollContext extends BaseContext implements IPollFunctions {
```
Comment (tomi) [inline packages/core/src/node-execution-context/poll-context.ts:37]: JSDocs would be nice

File: packages/workflow/src/Workflow.ts [hunk 4]
```diff
@@ -1402,14 +1402,18 @@ export class Workflow {
...
 					additionalData,
 					mode,
 					'manual',
 				);
-				return { data: await nodeType.poll.call(thisArgs) };
+				const data =
+					nodeType instanceof Node
+						? await nodeType.poll(context)
+						: await nodeType.poll.call(context);
+				return { data };
```
Comment (tomi) [inline packages/workflow/src/Workflow.ts:1416]: What's the reason for this change? Seems to be unrelated to PollContext? Type mismatch?

File: packages/core/src/node-execution-context/helpers/binary-helpers.ts
Comment (tomi) [inline packages/core/src/node-execution-context/helpers/binary-helpers.ts:1]: I assume these were copied 1:1 from `NodeExecuteFunctions`?

File: packages/core/src/node-execution-context/helpers/binary-helpers.ts [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+		return {
+			getBinaryPath: this.getBinaryPath.bind(this),
+			getBinaryMetadata: this.getBinaryMetadata.bind(this),
+			getBinaryStream: this.getBinaryStream.bind(this),
+			binaryToBuffer,
+			binaryToString,
+			prepareBinaryData: this.prepareBinaryData.bind(this),
+			setBinaryDataBuffer: this.setBinaryDataBuffer.bind(this),
+			copyBinaryFile: this.copyBinaryFile.bind(this),
+		};
```
Comment (tomi) [inline packages/core/src/node-execution-context/helpers/binary-helpers.ts:38]: Should we perhaps define all the methods on this class as properties instead so we wouldn't need to do the binding and create new function instances?

I.e.
```ts
...
getBinaryPath = (binaryDataId: string) => {
  ...
}
...
```

File: packages/core/src/node-execution-context/helpers/request-helpers.ts
Comment (tomi) [inline packages/core/src/node-execution-context/helpers/request-helpers.ts:1]: I assume these were copied 1:1 from `NodeExecuteFunctions`?

File: packages/core/src/node-execution-context/helpers/request-helpers.ts [hunk 6]
```diff
@@ -0,0 +1,381 @@
...
+		private readonly context: IAllExecuteFunctions,
+		private readonly workflow: Workflow,
+		private readonly node: INode,
+		private readonly additionalData: IWorkflowExecuteAdditionalData,
+		private readonly runExecutionData: IRunExecutionData | null = null,
+		private readonly connectionInputData: INodeExecutionData[] = [],
+	) {}
+
+	get exported(): RequestHelperFunctions {
+		return {
```
Comment (tomi) [inline packages/core/src/node-execution-context/helpers/request-helpers.ts:50]: Same comment here about should we avoid `.bind`

File: packages/core/src/node-execution-context/helpers/scheduling-helpers.ts [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+
+import { ScheduledTaskManager } from '@/ScheduledTaskManager';
+
+export class SchedulingHelpers {
+	private readonly scheduledTaskManager = Container.get(ScheduledTaskManager);
+
+	constructor(private readonly workflow: Workflow) {}
+
+	get exported(): SchedulingFunctions {
+		return {
```
Comment (tomi) [inline packages/core/src/node-execution-context/helpers/scheduling-helpers.ts:12]: And here as well

File: packages/workflow/src/Interfaces.ts [hunk 8]
```diff
@@ -985,8 +983,7 @@ export type IExecuteFunctions = ExecuteFunctions.GetNodeParameterFn &
 			BinaryHelperFunctions &
 			DeduplicationHelperFunctions &
 			FileSystemHelperFunctions &
-			SSHTunnelFunctions &
-			JsonHelperFunctions & {
+			SSHTunnelFunctions & {
```
Comment (tomi) [inline packages/workflow/src/Interfaces.ts:986]: Are we sure we are not breaking anything (e.g. custom nodes) by removing `JsonHelperFunctions`?

**Review by tomi — COMMENTED** [review]
Nice, starting to look much cleaner already 👌 Comments mainly about missing docs and how we expose the methods from helpers

File: packages/core/src/NodeExecuteFunctions.ts (hunk 1 above)
Comment (netroy) [reply packages/core/src/NodeExecuteFunctions.ts:170]: I haven't moved many functions out of `NodeExecutionFunctions` yet. Once I have extractd all the context objects, those functions will be moved to `BaseContext` as private methods. We can then remove this eslint exception.

File: packages/core/src/node-execution-context/base-contexts.ts (hunk 2 above)
Comment (netroy) [reply packages/core/src/node-execution-context/base-contexts.ts:15]: I want to document all the context objects together in a single PR, once I understand this code better. Created a [ticket](https://linear.app/n8n/issue/CAT-280) for that.

File: packages/core/src/node-execution-context/__tests__/base-context.test.ts [hunk 9]
```diff
@@ -0,0 +1,168 @@
...
+
+	beforeEach(() => {
+		jest.clearAllMocks();
+	});
+
+	describe('getNode', () => {
+		it('should return a deep copy of the node', () => {
+			const result = testContext.getNode();
+			expect(result).not.toBe(node);
+			expect(JSON.stringify(result)).toEqual(JSON.stringify(node));
```
[the remark this answers is not in the document]
Comment (netroy) [reply packages/core/src/node-execution-context/__tests__/base-context.test.ts:44]: for some weird reason `result` when logged to console is correct, but on `.toEqual` it's `undefined`. I couldn't wrap my head around what was happening because of all the proxy mock objects. I can clean this up later.

File: packages/core/src/node-execution-context/base-contexts.ts [hunk 10]
```diff
@@ -0,0 +1,107 @@
...
+
+	get logger() {
+		return LoggerProxy;
+	}
+
+	getExecutionId() {
+		return this.additionalData.executionId!;
+	}
+
+	getNode(): INode {
```
[the remark this answers is not in the document]
Comment (netroy) [reply packages/core/src/node-execution-context/base-contexts.ts:33]: Will do in the docs PR. So far, even I don't know why we do this. 
removing the `deepCopy` breaks too many tests, so I want to try making that change later as well.

File: packages/core/src/node-execution-context/base-contexts.ts [hunk 11]
```diff
@@ -0,0 +1,107 @@
...
+		return deepCopy(this.node);
+	}
+
+	getWorkflow() {
```
[the remark this answers is not in the document]
Comment (netroy) [reply packages/core/src/node-execution-context/base-contexts.ts:37]: same as above. I will likely change this method, and then document it accordingly then.

File: packages/workflow/src/Workflow.ts (hunk 4 above)
Comment (netroy) [reply packages/workflow/src/Workflow.ts:1416]: A long time ago I started the [initiative](https://github.com/n8n-io/n8n/pull/6464) to stop abusing `this` as the node execution context in the nodes. I want to pursue that again, so that we can reduce a lot of code duplication in the nodes.

This change doesn't need to be in this PR. I'll remove it for now.

File: packages/core/src/node-execution-context/helpers/binary-helpers.ts (hunk 5 above)
Comment (netroy) [reply packages/core/src/node-execution-context/helpers/binary-helpers.ts:38]: eventually, yes. but for now I want these PRs to make the least amount of changes possible.

File: packages/workflow/src/Interfaces.ts (hunk 8 above)
Comment (netroy) [reply packages/workflow/src/Interfaces.ts:986]: we did not remove it, we merged it in `BaseHelperFunctions`. This is all backward compatible.

File: packages/core/src/node-execution-context/__tests__/base-context.test.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (tomi) [reply packages/core/src/node-execution-context/__tests__/base-context.test.ts:44]: Not a biggie, this is completely fine as well 👍

File: packages/core/src/node-execution-context/helpers/binary-helpers.ts [hunk 12]
```diff
@@ -0,0 +1,148 @@
...
+import path from 'path';
+import type { Readable } from 'stream';
+import Container from 'typedi';
+
+import { BinaryDataService } from '@/BinaryData/BinaryData.service';
+import { binaryToBuffer } from '@/BinaryData/utils';
+// eslint-disable-next-line import/no-cycle
+import { binaryToString } from '@/NodeExecuteFunctions';
+
+export class BinaryHelpers {
```
Comment (ivov) [inline packages/core/src/node-execution-context/helpers/binary-helpers.ts:20]: Can we have `BinaryFunctions`, `SchedulingFunctions`, and `RequestFunctions`? And they return `ExportedBinaryFunctions`, `ExportedSchedulingFunctions`, `ExportedRequestFunctions`?

File: packages/core/src/node-execution-context/helpers/binary-helpers.ts (hunk 5 above)
Comment (netroy) [reply packages/core/src/node-execution-context/helpers/binary-helpers.ts:38]: We actually currently also create new function instances for execution context. And, `this.helpers` is (unfortunately) not the same object across various execution context types.
I'd like to merge all the helpers into a single file, and make all the same helpers available across all context types. 
Created https://linear.app/n8n/issue/CAT-281 to do that later.

File: packages/core/src/node-execution-context/base-contexts.ts [hunk 13]
```diff
@@ -0,0 +1,107 @@
...
+	Workflow,
+	WorkflowExecuteMode,
+} from 'n8n-workflow';
+import { deepCopy, LoggerProxy } from 'n8n-workflow';
+import { Container } from 'typedi';
+
+import { InstanceSettings } from '@/InstanceSettings';
+
+export abstract class BaseContext implements Omit<FunctionsBase, 'getCredentials'> {
+	protected readonly instanceSettings = Container.get(InstanceSettings);
```
[the remark this answers is not in the document]
Comment (netroy) [reply packages/core/src/node-execution-context/base-contexts.ts:16]: I wanted to do that. but `typedi` only does singletons, and we need a factory here.
I'd like to explore the option of replacing typedi with something custom, which also has a `Container.newInstance` method to be able to create multiple instances of any class using DI.

File: packages/core/src/node-execution-context/base-contexts.ts [hunk 14]
```diff
@@ -0,0 +1,107 @@
...
+	getNode(): INode {
+		return deepCopy(this.node);
+	}
+
+	getWorkflow() {
+		const { id, name, active } = this.workflow;
+		return { id, name, active };
+	}
+
+	getMode() {
```
[the remark this answers is not in the document]
Comment (netroy) [reply packages/core/src/node-execution-context/base-contexts.ts:42]: That would be a breaking change unfortunately.

File: packages/core/src/node-execution-context/base-contexts.ts [hunk 15]
```diff
@@ -0,0 +1,107 @@
...
+
+	getTimezone() {
+		return this.workflow.timezone;
+	}
+
+	getCredentialsProperties(type: string) {
+		return this.additionalData.credentialsHelper.getCredentialsProperties(type);
+	}
+
+	async prepareOutputData(outputData: INodeExecutionData[]) {
```
[the remark this answers is not in the document]
Comment (netroy) [reply packages/core/src/node-execution-context/base-contexts.ts:104]: not in our repo since a year. But custom node could still be using it. 
this was deprecated a while ago. but I also don't want to remove it in this PR.

File: packages/core/src/node-execution-context/base-contexts.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (netroy) [reply packages/core/src/node-execution-context/base-contexts.ts:15]: `BaseContext` is definitely not a great name, but `NodeExecutionContext` also feels confusing. perhaps we can rename this later?

[the remark this answers is not in the document]
Comment (tomi) [reply packages/core/src/node-execution-context/base-contexts.ts]: Since this is abstract class, maybe `BaseExecutionContext`?

**Review by tomi — APPROVED** [review]

**Comment by janober** [comment]
Got released with [`n8n@1.67.0`](https://github.com/n8n-io/n8n/releases/tag/n8n%401.67.0)
